include/net/net_timeout.h

```diff
@@ -0,0 +1,66 @@
+/** @file
+ * @brief Network timer with wrap around
+ *
+ * Timer that runs longer than about 49 days needs to calculate wraps.
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __NET_TIMEOUT_H
+#define __NET_TIMEOUT_H
+
+/**
+ * @brief Network long timeout primitives and helpers
+ * @defgroup net_timeout Network long timeout primitives and helpers
+ * @ingroup networking
+ * @{
+ */
+
+#include <string.h>
+#include <zephyr/types.h>
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Let the max timeout be 100 ms lower because of
+ * possible rouding in delayed work implementation.
+ */
+#define NET_TIMEOUT_MAX_VALUE ((u32_t)(INT32_MAX - 100))
+
+struct net_timeout {
+	/** Used to track timers */
+	sys_snode_t node;
+
+	/** Address lifetime timer start time */
+	u32_t timer_start;
+
+	/** Address lifetime timer timeout in milliseconds. Note that this
+	 * value is signed as k_delayed_work_submit() only supports signed
+	 * delay value.
+	 */
+	s32_t timer_timeout;
+
+	/** Timer wrap count. Used if the timer timeout is larger than
+	 * about 24 days. The reason we need to track wrap arounds, is
+	 * that the timer timeout used in k_delayed_work_submit() is
+	 * 32-bit signed value and the resolution is 1ms.
+	 */
+	s32_t wrap_counter;
```

Comment:
But why? Why do we need longer address lifetimes than 24 days? What's wrong with clamping it down to 24 days and refreshing then? It's not worse than what happens when device reboots and has to request an address again, is it?

The whole point of calling for 32-but start time and delays was to have simple, efficient code. And now with (INT32_MAX - 100) and 64-bit devision it's back to dirty bloaty magic.

Reply:
Because you can have longer lifetime than 24 days, thus we should be able to handle that number. For this we definitely need wrap support.

Reply:
> Because you can have longer lifetime than 24 days, thus we should be able to handle that number. For this we definitely need wrap support.

Sorry, but to handle "longer lifetime than 24 days", really handle, I mean, It's not enough to just count longer than 24 days. You need to store the addresses in non-volatile memory, and have real-time clock mandatory, so after reboot you could check if it's still alive. For example, if you got an address for 100 days, rebooted next day, then you should not request an address, but use one you stored in flash for next 99 days still.

Do you do anything like that? No, doesn't make sense. The same, it doesn't make sense to go out of the way to support 100-days addresses at all. Just round it down to 24 days, and re-request after that. No animals will be harmed by doing so.

Does anybody really need 100-day addresses? Let them stand up and raise the hand, so we can put names to the requirements. In this case, no problem either - introduce concept of "short" network timers and "long" ones. Short are of course 32-bit. Long by default are the same as short, but there's an option to make them 64bit - fully 64 bit, 64-bit start time, 64-bit delay. No need to pseudo-optimize memory usage - whoever needs memory won't enable 64-bit "long" timers in the firsts place.

So, confusing "wrap count" support is not needed, whether we target for optimal memory usage (all timers 32 bit), or to follow the word of standard with higher memory overhead.

Reply:
> Sorry, but to handle "longer lifetime than 24 days", really handle, I mean, It's not enough to just count longer than 24 days. You need to store the addresses in non-volatile memory, and have real-time clock mandatory, so after reboot you could check if it's still alive.

There is no requirement to store anything like this to non-volatile memory. Why do you start to discuss something like this here.

Reply:
> There is no requirement to store anything like this to non-volatile memory.

Then maybe there's no requirement of supporting lifetimes longer than 24 days either? Or of it is, it can be separated into separate config?

> Why do you start to discuss something like this here.

I'm just trying to understand how after my comments in #9508 and your desire to rework it, this new code turned up - because it's completely not what I suggested. Maybe the simple answer is that you never wanted to rework it per my suggestions (you never said you would, indeed). (But then why to rework at all, you could push with #9508, you definitely can push any code you like.)

Reply:
> You are mistaken here, there needs to be some kind of consensus about the patches, not all need to agree with the patches thou.

So, that's why I write all what I write - to have better discussion (even if it's post-discussion, sadly!), and hopefully consensus. And IMHO, we still don't have enough discussions. Monthly "forum" meetings a great for general direction, but details of how stuff should be done are oftentimes skipped, "implied" (meaning that different parties can have different ideas of how it should be done), etc.

> About this PR vs the earlier version in #9508, this one tries to lower the memory more by not using 64-bit timeout values. There is no functionality changes between those two, only implementation is different.

Not exactly. The current master has:

~~~
        /** Address lifetime timer start time */
        s64_t lifetime_timer_start;

        /** lifetime timer timeout */
        u32_t lifetime_timer_timeout;
~~~

#9508 had:

~~~
u64_t timer_start;
u32_t timer_timeout;
~~~

This PR has:

~~~
u32_t timer_start;
s32_t timer_timeout;
s32_t wrap_counter;
~~~

What I proposed:

~~~
u32_t timer_start;
u32_t timer_timeout;
~~~

So, you didn't improve memory usage at all. Yes, with this PR you added new feature - support for lifetimes over 24 days. But why you went for that at all? This work was supposed to be a "cleanup/optimize" work based on https://github.com/zephyrproject-rtos/zephyr/issues/8728 , not adding new features.

Reply:
> u32_t lifetime_timer_timeout;

That was actually wrong, we would not be able to have >24 day limit with that. Thus it is pointless to try to compare with that solution.

Reply:
> So, you didn't improve memory usage at all

I definitely did save space. I just wonder did you read the PR and understood what it is trying to do.

Reply:
>>     u32_t lifetime_timer_timeout;

> That was actually wrong, we would not be able to have >24 day limit with that.

That's the solution in the master. We **were not** (not "would not") able to have 24 day limit, and nobody cared. You were supposed (unless I'm mistaken) to do functionality-invariant cleanup/optimization and reduce technical debt. Instead, you "improved" functionality (by supporting >24 days) and increased technical debt (because there's duplicated code which can/should be refactored, and you say "it can be done later"). That's how I see it, sorry.

Of course, you did other things as "was intended" (even per my outlook), like `-struct k_delayed_work lifetime;/+	struct net_timeout lifetime;`, nobody challenges that.